src/main/java/com/direwolf20/buildinggadgets/common/util/blocks/RegionSnapshot.java

```diff
@@ -0,0 +1,361 @@
+                }
+            }
+            // Palette serialization ends
+
+            // Hence we record the storage frame after it is completed, the initial values of the counter would be used as a storage frame regardless, therefore we need to skip inorder for this algorithm to function correctly
+            if (firstIteration)
+                firstIteration = false;
+            else {
+                // If the same block state showed up again, increase the number of streaks
+                // If we have more than 255 streaks (maximum integer that can be stored in 8 bits), we force start a new streak, because the current storage frame won't fit that many of them
+                if (state.equals(lastStreak) && streak < 255) {
+                    streak++;
+                    continue;
+                }
+
+                // Note: if the program reached here, it means the current storage frame is completed
+
+                // - Number of streaks and the ID of the block state that is streaking
+                // - No need to differentiate between single and repeating block states, since the number at the larger endian represents how many repeating block states are there
+                // - If the streaking block state is empty, we use the preoccupied ID 0
+                int frame = ((streak & 0xff) << 24) | (mapPalettes.getOrDefault(lastStreak, 0) & 0xffffff);
+                frames.add(frame);
+            }
+
+            // Regardless of which situation, we reset the counter
+            streak = 1;
+            lastStreak = state;
+        }
+
+        tag.setIntArray(BLOCK_FRAMES, frames.toIntArray());
+        tag.setTag(BLOCK_PALETTES, palettes);
+
+        NBTTagList tileData = new NBTTagList();
+        for (Pair<BlockPos, NBTTagCompound> data : snapshot.tileData) {
+            NBTTagCompound serializedData = new NBTTagCompound();
+            serializedData.setTag(TILE_POS, NBTUtil.writeBlockPos(data.getLeft()));
+            serializedData.setTag(TILE_NBT, data.getRight());
+            tileData.add(serializedData);
+        }
+        tag.setTag(TILE_DATA, tileData);
+
+        return tag;
+    }
+
+    private static RegionSnapshot deserialize(NBTTagCompound tag) {
+        int dimension = tag.getInt(DIMENSION);
+        World world = ServerLifecycleHooks.getCurrentServer().getWorld(Objects.requireNonNull(DimensionType.getById(dimension)));
+
+        Region region = Region.deserializeFrom(tag);
+
+        List<IBlockState> palettes = new ArrayList<>();
+        // "nothing should be here" block state, represented with 'null'
+        palettes.add(null);
+        NBTTagList palettesNBT = tag.getList(BLOCK_PALETTES, Constants.NBT.TAG_COMPOUND);
+        for (int i = 0; i < palettesNBT.size(); i++) {
+            palettes.add(NBTUtil.readBlockState(palettesNBT.getCompound(i)));
+        }
+        assert palettes.size() == palettesNBT.size() + 1;
```

Comment:
Not really, because we build `palettes` during the iterating process. Technically this assertion is necessary, because we know
1. `for(int i = 0; i < array.length; i++)` is reliable
2. `palettes.add(Optional.empty())` is reliable

Reply:
That's not what I meant... I thought you intended to say with this, that there is something to iterate on.

Well, if you want to do it that way, ok.